### PR TITLE
Add monitored-entity reconciler to logmodule reconciler

### PR DESCRIPTION
**Pull request by 0sewa0**
## Description
[DAQ-1397](https://dt-rnd.atlassian.net/browse/DAQ-1397)

Both `KubernetesClusterMEID` and `KubernetesClusterName` are a hard requirement for the deployment of the standalone logmodule.

So we add the `monitoredentities.Reconciler` to the logmodule reconcile, the ensure that we have it available.
- If we still don't have the `KubernetesClusterMEID` and `KubernetesClusterName` then we log it and then end the logmodule reconcile early.
   - should we error instead of log? 

## How can this be tested?

Deploy a dynakube with `logModule.enable: true` in 2 scenarios:
1. When there is an ME on your tenant => everything should deploy fine (daemonset appears)
2. When there is **NO** ME on your tenant => should see in the logs that something is not good (daemonset does NOT appear)

[DAQ-1397]: https://dt-rnd.atlassian.net/browse/DAQ-1397?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: pkg/controllers/dynakube/logmodule/reconciler.go [hunk 1]
```diff
@@ -38,11 +45,23 @@ func NewReconciler(clt client.Client,
...
 }
 
 func (r *Reconciler) Reconcile(ctx context.Context) error {
-	err := r.oneAgentConnectionInfoReconciler.Reconcile(ctx)
+	err := r.monitoredEntitiesReconciler.Reconcile(ctx)
+	if err != nil {
+		return err
+	}
+
+	if r.dk.Status.KubernetesClusterMEID == "" || r.dk.Status.KubernetesClusterName == "" {
```
Comment (waodim) [inline pkg/controllers/dynakube/logmodule/reconciler.go:58]: What do you think about abstracting this? either both fields abstract on their own or abstract the whole check? Or do you think this is too much?

File: pkg/controllers/dynakube/logmodule/reconciler.go [hunk 2]
```diff
@@ -38,11 +45,23 @@ func NewReconciler(clt client.Client,
...
+		log.Info("the status of the DynaKube is missing information about the kubernetes monitored-entity, skipping logmodule deployment")
```
Comment (waodim) [inline pkg/controllers/dynakube/logmodule/reconciler.go:59]: I personally do not mind by not erroring here and instead just log it and return early.

Do you think we could add this case somehow to the dynakube status with an appropriate reason?

Comment (chrismuellner) [reply pkg/controllers/dynakube/logmodule/reconciler.go:59]: we should add this to the deployment status of the log module so it's obvious not only in logs but also in the Dynakube itself

**Review by aorcholski — COMMENTED** [review]
1. I created a cluster from scratch (to get rid of ClusterMEID)
2. I applied dynakube
```
spec:                                                                                                                                                                                                                                                                                                                                                                                         
  apiUrl: https://<TENANT>.dev.dynatracelabs.com/api                                                                                                                                                                                                                                                                                                                                          
  logModule:                                                                                                                                                                                                                                                                                                                                                                                  
    enabled: true                                                                                                                                                                                                                                                                                                                                                                             
  templates:                                                                                                                                                                                                                                                                                                                                                                                  
    logModule:                                                                                                                                                                                                                                                                                                                                                                                
      imageRef:                                                                                                                                                                                                                                                                                                                                                                               
        repository: .....
        tag: latest 
```
3. Operator is RUNNING
```
NAME                              APIURL                                       STATUS    AGE
dynakube.dynatrace.com/dynakube   https://tec31546.dev.dynatracelabs.com/api   Running   10m
```
status :
```
  conditions:
  - lastTransitionTime: "2024-10-16T15:33:11Z"
    message: Kubernetes Cluster MEID is outdated in the status
    reason: StatusOutdated
    status: "False"
    type: MonitoredEntity
```
logs :
```
{"level":"info","ts":"2024-10-16T15:40:12.758Z","logger":"monitored-entities","msg":"start reconciling monitored entities"}
{"level":"info","ts":"2024-10-16T15:40:12.795Z","logger":"monitored-entities","msg":"no MEs found, no kubernetesClusterMEID will be set in the dynakube status"}
```
but NOTHING about logModule
```
[]$ kctl logs deployment.apps/dynatrace-operator | grep "the status of the DynaKube is missing information about the kubernetes monitored-entity"
[]$ 
```
I know everything is "fine" in this case but it looks like a bug - dynakube is RUNNING but where is my logModule ?

**Comment by 0sewa0** [comment]
@aorcholski I didn't to reproduce that, I get:

```
{"level":"info","ts":"2024-10-17T07:29:46.687Z","logger":"logmodule-daemonset","msg":"the status of the DynaKube is missing information about the kubernetes monitored-entity, skipping logmodule deployment"}
```

But yes the `Phase` of the DynaKube is `Running`, but that was semi intended. My question of `should we error instead of log?` is related to that. Will try to see what happens if we just error.

File: pkg/controllers/dynakube/logmodule/reconciler.go (hunk 2 above)
Comment (0sewa0) [reply pkg/controllers/dynakube/logmodule/reconciler.go]: I now error instead, and you will see in the logs: (it's a bit loud)
```
{"level":"error","ts":"2024-10-17T07:51:01.989Z","logger":"dynakube","msg":"error reconciling DynaKube","namespace":"dynatrace","name":"zib50933zib50933zib50933zib50933zib5093","error":"the status of the DynaKube is missing information about the kubernetes monitored-entity, skipping logmodule deployment","stacktrace":"github.com/Dynatrace/dynatrace-operator/pkg/controllers/dynakube.(*Controller).handleError\n\tgithub.com/Dynatrace/dynatrace-operator/pkg/controllers/dynakube/controller.go:191\ngithub.com/Dynatrace/dynatrace-operator/pkg/controllers/dynakube.(*Controller).Reconcile\n\tgithub.com/Dynatrace/dynatrace-operator/pkg/controllers/dynakube/controller.go:145\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).Reconcile\n\tsigs.k8s.io/controller-runtime@v0.19.0/pkg/internal/controller/controller.go:116\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).reconcileHandler\n\tsigs.k8s.io/controller-runtime@v0.19.0/pkg/internal/controller/controller.go:303\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).processNextWorkItem\n\tsigs.k8s.io/controller-runtime@v0.19.0/pkg/internal/controller/controller.go:263\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).Start.func2.2\n\tsigs.k8s.io/controller-runtime@v0.19.0/pkg/internal/controller/controller.go:224"}
{"level":"info","ts":"2024-10-17T07:51:01.989Z","logger":"dynakube","msg":"reconciling DynaKube finished","namespace":"dynatrace","name":"zib50933zib50933zib50933zib50933zib5093","result":{"Requeue":false,"RequeueAfter":0}}
{"level":"error","ts":"2024-10-17T07:51:01.990Z","logger":"main","msg":"Reconciler error","controller":"dynakube-controller","controllerGroup":"dynatrace.com","controllerKind":"DynaKube","DynaKube":{"name":"zib50933zib50933zib50933zib50933zib5093","namespace":"dynatrace"},"namespace":"dynatrace","name":"zib50933zib50933zib50933zib50933zib5093","reconcileID":"cc5891b6-94e1-4c21-8298-0c8b432ded14","error":"the status of the DynaKube is missing information about the kubernetes monitored-entity, skipping logmodule deployment","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).reconcileHandler\n\tsigs.k8s.io/controller-runtime@v0.19.0/pkg/internal/controller/controller.go:316\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).processNextWorkItem\n\tsigs.k8s.io/controller-runtime@v0.19.0/pkg/internal/controller/controller.go:263\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).Start.func2.2\n\tsigs.k8s.io/controller-runtime@v0.19.0/pkg/internal/controller/controller.go:224"}

```

Also the `Phase` of the `DynaKube` will be in `Error`

Is this what we want? (I kinda want this 😅 )

**Review by aorcholski — COMMENTED** [review]
I have reinstalled operator and dynakube from scratch (twice)
```
NAME                              APIURL                                       STATUS    AGE
dynakube.dynatrace.com/dynakube   https://tec31546.dev.dynatracelabs.com/api   Running   31m
```
no error, no logmodule pod...

**Review by aorcholski — COMMENTED** [review]
**EXISTING** namespace "disables" logmodule ERROR
```
spec:                                                                                                                                                                                                                                                                                                                                                                                         
  apiUrl: https://TENANT.dev.dynatracelabs.com/api                                                                                                                                                                                                                                                                                                                                          
  dynatraceApiRequestThreshold: 15                                                                                                                                                                                                                                                                                                                                                            
  logModule:                                                                                                                                                                                                                                                                                                                                                                                  
    enabled: true                                                                                                                                                                                                                                                                                                                                                                             
  networkZone: nztest                                                                                                                                                                                                                                                                                                                                                                         
```

**Review by luhi-DT — APPROVED** [review]
lgtm

**Review by waodim — APPROVED** [review]
tested it - worked for me (I manually deleted the MEID with a curl command from the status) code also lgtm

**Pull request merged**